docs/user-guide/optimization.md

```diff
@@ -0,0 +1,120 @@
+
+Now we call `visualize` again, this time not setting `optimize_graph` so it picks up its default value of `True`.
+
+```python
+c.visualize()
+```
+
+![Map fusion optimization](../images/optimization_map_fusion.svg)
+
+Note that with optimization turned on, the array `b` is no longer written as an intermediate output since it will be computed in the same tasks that compute array `c`. The overall number of tasks is reduced from 10 to 5, and the intermediate data (total `nbytes`) is reduced too.
+
+Here we have just called `visualize` with the `optimize_graph` argument, but it's possible to use it when calling `compute` - which can be useful when debugging a computation.
+
+```python
+c.compute(optimize_graph=False)
+```
+
+## Multiple-input fusion
+
+Cubed supports more powerful optimizations, such as for when an array is created from multiple input arrays. Here is an example, shown first with optimization turned off.
+
+```python
+import cubed.array_api as xp
+
+a = xp.ones((3, 3), chunks=(2, 2))
+b = xp.ones((3, 3), chunks=(2, 2))
+c = xp.ones((3, 3), chunks=(2, 2))
+d = xp.add(b, c)
+e = xp.add(a, d)
+
+e.visualize("cubed-unoptimized", optimize_graph=False)
+```
+
+![Multiple inputs unoptimized](../images/optimization_multiple_inputs_unoptimized.svg)
+
+And with optimization turned on (the default):
+
+```python
+e.visualize()
+```
+
+![Multiple inputs optimized](../images/optimization_multiple_inputs.svg)
+
+Notice how the array `d` is fused away.
+
+## Debugging optimization
+
+Sometimes it can be difficult to understand why particular operations in a computation plan have been fused together - or more commonly, why they have *not* been fused. By enabling debug logging you can get detailed information from the optimize function to help you understand which operations are being fused - or not - and the reason in either case.
+
+Here's the previous example with logging enabled:
+
+```python
+import logging
+
+logging.basicConfig(level=logging.DEBUG)
+
+e.visualize()
+```
+
+The output explains which operations can or can't be fused, and why:
+
+```
+DEBUG:cubed.core.optimization:can't fuse op-001 since it is not a primitive operation, or it uses map_direct
+DEBUG:cubed.core.optimization:can't fuse op-002 since it is not a primitive operation, or it uses map_direct
+DEBUG:cubed.core.optimization:can't fuse op-003 since it is not a primitive operation, or it uses map_direct
+DEBUG:cubed.core.optimization:can't fuse op-004 since no predecessor ops can be fused
+DEBUG:cubed.primitive.blockwise:can fuse op-005 since num tasks of predecessor ops match
+```
+
+## Advanced settings
+
+There are limits to how many input arrays and input chunks reads are fused together. These are imposed so that the number of reads that an individual task must perform is not excessive, which would otherwise result in slow running tasks.
+
+In some cases you may want to change these limits, which we look at here.
```

Comment:
Do we want to explain what some of the cases might be? Or leave that as future work?

Reply:
It could certainly be expanded, but I'd prefer to leave it for now.